File: server/sound/SoundCloud.ts
import fs from "fs";
import { join } from "path";
import { SoundcloudTrackV2 } from "soundcloud.ts";
import { exec } from "youtube-dl-exec";
import { sanitizeFileName } from "../MediaManager";
import { Media, SoundCloud } from "../server";
import { Downloadable } from "../struct";

export async function searchSoundCloud(query: string): Promise<Downloadable[]> {
  try {
    const results =
      (await SoundCloud.api.tracks.searchV2({ q: query, limit: 30 }))?.collection.filter(
        (a) => a
      ) || [];
    return results.map(constructVideoFromSoundCloud);
  } catch (err) {
    console.error(`Error searching soundcloud: ${err}\nQuery: '${query}'\n${err.stack}`);
    return [];
  }
}

export async function downloadSoundcloud(url: string): Promise<{ err: string } | string> {
  return new Promise(async (res) => {
    try {
      const path = join(Media.tempdir(), `${sanitizeFileName(url)}.mp3`);
      const file = fs.createWriteStream(path);
      const stream = exec(url, {
        format: "bestaudio",
        output: "-",
      });
      stream.stdout.pipe(file);
      file.on("error", (err) => {
        res({ err: String(err) });
      });
      file.on("finish", () => {
        res(path);
      });
    } catch (err) {
      res({ err });
    }
  });
}

export function constructVideoFromSoundCloud(track: SoundcloudTrackV2): Downloadable {
  return {
    title: track.title || "",
    duration: track.full_duration || track.duration || 0,
    uploadedAt: new Date(track.created_at).getTime(),
    views: track.playback_count || 0,
    thumbnail: track.artwork_url || track.user.avatar_url || "",
    author: {
      name: track.user.username || "",
      icon: track.user.avatar_url || "",
      url: track.user.permalink_url || "",
    },
    url: track.permalink_url || "",
    embed: `https://w.soundcloud.com/player/?url=https%3A//api.soundcloud.com/tracks/${track.id}&color=%23ff5500&auto_play=false&hide_related=false&show_comments=true&show_user=true&show_reposts=false&show_teaser=true&visual=true`,
  };
}

File: server/sound/soundcloud.go
package sound

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"os"
	"path"
	"time"

	"github.com/itzTheMeow/YTMusic/metadata"
	"github.com/itzTheMeow/YTMusic/types"
	"github.com/itzTheMeow/YTMusic/util"
	soundcloud "github.com/juanefec/soundcloud-api"
	"github.com/oklog/ulid/v2"
)

func SearchSoundCloud(query string) []types.Downloadable {
	list := make([]types.Downloadable, 0)
	if metadata.SoundCloudClient == nil {
		return list
	}
	res, err := metadata.SoundCloudClient.Search(soundcloud.SearchOptions{
		Query: query,
		Limit: 12,
		Kind:  soundcloud.KindTrack,
	})
	if err == nil {
		for _, item := range res.Collection {
			track := soundcloud.Track{}
			b, err := json.Marshal(item)
			if err != nil {
				continue
			}
			err = json.Unmarshal(b, &track)
			if err != nil {
				continue
			}
			list = append(list, ConstructDownloadableFromSoundCloud(track))
		}
	} else {
		log.Println(err)
	}
	return list
}

func ConstructDownloadableFromSoundCloud(track soundcloud.Track) types.Downloadable {
	thumb := track.ArtworkURL
	if thumb == "" {
		thumb = track.User.AvatarURL
	}
	return types.Downloadable{
		Title:      track.Title,
		Duration:   int(track.FullDurationMS),
		UploadedAt: int(util.Grab(time.Parse("2006-01-02T15:04:05Z", track.CreatedAt)).UnixMilli()),
		Views:      int(track.PlaybackCount),
		Thumbnail:  thumb,
		Author: types.DownloadableAuthor{
			Name: track.User.Username,
			Icon: track.User.AvatarURL,
			URL:  track.User.PermalinkURL,
		},
		URL:   track.PermalinkURL,
		Embed: fmt.Sprintf(`https://w.soundcloud.com/player/?url=https%%3A//api.soundcloud.com/tracks/%v&color=%%23ff5500&auto_play=false&hide_related=false&show_comments=true&show_user=true&show_reposts=false&show_teaser=true&visual=true`, track.ID),
	}
}

func DownloadSoundcloud(url string) (*string, error) {
	filepath := path.Join(os.TempDir(), ulid.Make().String())

	result, err := DownloadURL(url)
	if err != nil {
		return nil, err
	}
	file, err := os.Create(filepath)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	written, err := io.Copy(file, result)
	if err != nil {
		return nil, err
	}
	if written == 0 {
		return nil, errors.New("Failed to download track.")
	}
	return &filepath, nil
}
